### ISSUE TITLE
In k8s cluster, linkerd-identity may get descheduled when scaling down the cluster

### ISSUE THREAD

**Issue by pengux** [linkerd/linkerd2#5356]
### What is the issue?
As there is no `priorityClassName` set for the `linkerd-identity` deployment, it may get descheduled completed if the cluster is scaled down and worker nodes are removed. This will prevent the `linkerd-proxy` from working when the cluster is scaled back up again.

### How can it be reproduced?
1. Install `linkerd2` in a Kubernetes cluster with 3 worker nodes with the High Availability configuration. There should be 2 pods for the `linkerd-identity` service
2. Scale down the cluster to only 1 worker node, this should max out the nodes resources so that it must not be able to schedule a `linkerd-identity` on that node. This may need to be down several times until the `linkerd-identity` is complete descheduled in the cluster.
3. Scale the cluster back to 3 worker nodes, now the `linkerd-proxy` containers will fail to be in a `Ready` state.

### Environment

- Kubernetes Version: 1.18
- Cluster Environment: EKS
- Host OS:
- Linkerd version: 2.9.0

### Possible solution
Allow the priorityClassName of the `linkerd-identity` deployment to be set, so the pod doesn't get completely descheduled in the cluster when scaling down.

**Comment by adleong** [comment]
Thanks for the PR, @pengux!  Before I review that, I wanted to ask some higher level questions here about the issue itself.

Linkerd does not support scaling clusters down below 3 nodes when in HA mode.  We cannot provide HA guarantees for clusters of that size and so HA mode is not appropriate for any cluster which may get scaled down below 3 nodes.  What exactly is your use-case here?

**Comment by pengux** [comment]
> Thanks for the PR, @pengux! Before I review that, I wanted to ask some higher level questions here about the issue itself.
> 
> Linkerd does not support scaling clusters down below 3 nodes when in HA mode. We cannot provide HA guarantees for clusters of that size and so HA mode is not appropriate for any cluster which may get scaled down below 3 nodes. What exactly is your use-case here?

Our use-case is to be able to scale down dev clusters to save cost, also being able to recycle nodes. From the slack thread, we already know that this is not currently supported in Linkerd2 yet. This change is one step to support it though.

**Comment by adleong** [comment]
Does this happen when HA mode is not enabled, or only with the HA flag?  I would say that HA mode should definitely be disabled for any cluster that you plan to scale down.  If this issue happens for clusters which are not in HA mode, more investigation is required to find out why this is happening.  I would expect that even without setting a priority class, the linkerd-identity deployment should be able to be rescheduled once resources become available.

**Comment by leghissa** [comment]
This specific issue is independent from HA mode, it was happening also when we had HA mode disabled. 
The problem is that when we scale down to 1 node the resources on that node are not enough to successfully schedule all the pods(we don't mind because nobody is going to actually use the cluster). And given that the `linkerd-identity` pod has priority 0 it doesn't get rescheduled.
The idea was to increase priority so that at least one `linkerd-identity` would always be running.

**Comment by adleong** [comment]
Interesting!  I would expect that once the cluster is scaled up again, that the `linkerd-identity` pod would get scheduled when resources become available and the cluster should resume working (though perhaps any Linkerd pods which got scheduled while the `linkerd-identity` pod was unavailable might need to be restarted).  Let me try this and get back to you.

**Comment by adleong** [comment]
I tested this by first scaling the `linkerd-identity` deployment down to 0 replicas.  Then I attempted to deploy the emojivoto app, injected with the proxy.  

As expected, the Linkerd proxy container in the emojivoto pods do not become ready.  The proxy logs contain this message over and over:

```
[    13.134486s]  WARN ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}:accept{peer.addr=10.244.0.159:54200 target.addr=10.244.0.223:4191}: rustls::session: Sending fatal alert AccessDenied
[    13.134648s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}:accept{peer.addr=10.244.0.159:54200 target.addr=10.244.0.223:4191}: linkerd2_app_core::serve: Connection closed error=unexpected error: no server certificate chain resolved
```

I then scaled the `linkerd-identity` deployment back up to 1 replicas.  Shortly after, this message appeared in the Linkerd proxy logs and the container became ready:

```
[   344.019638s]  INFO ThreadId(02) daemon:identity: linkerd2_app: Certified identity: web.emojivoto.serviceaccount.identity.linkerd.cluster.local
[   360.967932s]  INFO ThreadId(01) inbound:accept{peer.addr=10.244.0.166:39364 target.addr=10.244.0.223:8080}: linkerd2_app_core::serve: Connection closed error=connection error: Broken pipe (os error 32)
```

In other words, the cluster recovered as I would expect once the `linkerd-identity` controller became available again.  Does this not match what you are seeing?

**Comment by adleong** [comment]
Closing this due to inactivity.  Please re-open if you can provide the requested information.
